storage/src/mmr/verification.rs

```diff
@@ -318,6 +319,57 @@ impl<D: Digest> Proof<D> {
             digests,
         })
     }
+
+    /// Extract the hashes of all nodes that should be pinned at the given pruning boundary
+    /// from a proof that proves a range starting at that boundary.
+    ///
+    /// # Arguments
+    /// * `start_element_pos` - Start of the proven range (must equal pruning_boundary)
+    /// * `end_element_pos` - End of the proven range
+    ///
+    /// # Returns
+    /// A Vec of digest values for all nodes in `nodes_to_pin(pruning_boundary)`,
+    /// in the same order as returned by `nodes_to_pin` (decreasing height order)
+    pub fn extract_pinned_nodes(
+        &self,
+        start_element_pos: u64,
+        end_element_pos: u64,
+    ) -> Result<Vec<D>, Error> {
+        // Get the positions of all nodes that should be pinned
+        let pinned_positions: Vec<u64> = Self::nodes_to_pin(start_element_pos).collect();
+
+        // Get all positions required for the proof
+        let required_positions =
+            Self::nodes_required_for_range_proof(self.size, start_element_pos, end_element_pos);
+
+        // Should be the nodes in the proof
+        if required_positions.len() != self.digests.len() {
+            debug!(
+                "Proof digest count ({}) doesn't match required positions ({})",
+                self.digests.len(),
+                required_positions.len()
```

Comment:
Using structured logging here instead of format strings would make Patrick happy.